### PR TITLE
some fork() reworking

### PR DESCRIPTION
**Pull request by poettering**
We keep repeating very similar code over and over again, let's unify this. And other fixes.

The patch set consists of fixes and refactoring only. Nothing new really added.

### CODE REVIEW

**Comment by poettering** [comment]
`KVM: entry failed, hardware error 0x0`

**Comment by yuwata** [comment]
With this, PAM related error occurs. When I login my system with normal user, then I got the following.
```
Dec 24 23:48:57 systemd-logind[368]: New session 9 of user watanabe.
Dec 24 23:48:57 systemd[2805]: pam_unix(systemd-user:session): session opened for user watanabe by (uid=0)
Dec 24 23:48:57 systemd[2805]: Sucessfully forked off '(sd-pam)' as PID 2808.
Dec 24 23:48:57 systemd[2805]: PAM initialization failed
Dec 24 23:48:57 systemd[2808]: Assertion 'close_nointr(fd) != -EBADF' failed at ../src/basic/fd-util.c:81, function safe_close(). Aborting.
Dec 24 23:48:57 systemd[2805]: user@1000.service: Executing: /usr/lib/systemd/systemd --user
```
And, when I login by root, then got segfault.
```
Dec 24 23:58:25 systemd[4917]: pam_unix(systemd-user:session): session opened for user root by (uid=0)
Dec 24 23:58:25 systemd-logind[368]: New session 14 of user root.
Dec 24 23:58:25 systemd[4917]: Sucessfully forked off '(sd-pam)' as PID 4920.
Dec 24 23:58:25 systemd[4920]: Assertion 'close_nointr(fd) != -EBADF' failed at ../src/basic/fd-util.c:81, function safe_close(). Aborting.
Dec 24 23:58:25 systemd[4917]: PAM initialization failed
Dec 24 23:58:25 systemd[4917]: user@0.service: Executing: /usr/lib/systemd/systemd --user
Dec 24 23:58:25 login[2793]: pam_unix(login:session): session opened for user root by LOGIN(uid=0)
Dec 24 23:58:25 login[2793]: ROOT LOGIN ON tty2
Dec 24 23:58:25 systemd[4923]: systemd-coredump@1-4921-0.service: Executing: /usr/lib/systemd/systemd-coredump
Dec 24 23:58:25 systemd-coredump[4923]: Process 4920 ((sd-pam)) of user 0 dumped core.
                                        
                                        Stack trace of thread 4920:
                                        #0  0x00007f3b3c09b66b raise (libc.so.6)
                                        #1  0x00007f3b3c09d381 abort (libc.so.6)
                                        #2  0x00007f3b3bc99d46 log_assert_failed_realm (libsystemd-shared-236.so)
                                        #3  0x00007f3b3bcf8171 safe_close (libsystemd-shared-236.so)
                                        #4  0x00007f3b3bd45f7b barrier_set_role (libsystemd-shared-236.so)
                                        #5  0x00005564ecb8a5da setup_pam.lto_priv.110 (systemd)
                                        #6  0x00005564ecc2c581 exec_child.constprop.10 (systemd)
                                        #7  0x00005564ecbca79d exec_spawn (systemd)
                                        #8  0x00005564ecbf7eae service_spawn (systemd)
                                        #9  0x00005564ecbf9f38 service_enter_start (systemd)
                                        #10 0x00005564ecbfa3a5 service_start.lto_priv.171 (systemd)
                                        #11 0x00005564ecbc59b2 job_perform_on_unit.lto_priv.489 (systemd)
                                        #12 0x00005564ecbec21a manager_dispatch_run_queue.lto_priv.494 (systemd)
                                        #13 0x00007f3b3bc9682a source_dispatch (libsystemd-shared-236.so)
                                        #14 0x00007f3b3bc96b1a sd_event_dispatch (libsystemd-shared-236.so)
                                        #15 0x00007f3b3bc96c97 sd_event_run (libsystemd-shared-236.so)
                                        #16 0x00005564ecc17f56 invoke_main_loop (systemd)
                                        #17 0x00005564ecb91c1e main (systemd)
                                        #18 0x00007f3b3c08500a __libc_start_main (libc.so.6)
                                        #19 0x00005564ecb92daa _start (systemd)
Dec 24 23:58:25 systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 86 (stored), closing.
```

File: src/core/execute.c [hunk 1]
```diff
@@ -1207,28 +1207,16 @@ static int setup_pam(
 
         parent_pid = getpid_cached();
 
-        pam_pid = fork();
-        if (pam_pid < 0) {
-                r = -errno;
+        r = safe_fork_full("(sd-pam)", fds, n_fds, FORK_CLOSE_ALL_FDS, &pam_pid);
```
Comment (yuwata) [inline src/core/execute.c]: `FORK_CLOSE_ALL_FDS` also closes barrier?

File: src/core/execute.c [hunk 2]
```diff
@@ -1207,28 +1207,16 @@ static int setup_pam(
...
                 goto fail;
-        }
-
-        if (pam_pid == 0) {
+        if (r == 0) {
                 int sig, ret = EXIT_PAM;
 
                 /* The child's job is to reset the PAM session on
                  * termination */
                 barrier_set_role(&barrier, BARRIER_CHILD);
```
Comment (yuwata) [inline src/core/execute.c:1218]: So, the barrier is already closed here?

**Comment by yuwata** [comment]
I've tested this with https://github.com/yuwata/systemd/commit/51819aa970b3e2c71d3be35e4c2b81549be3b32d, then the above sd-pam related error disappears.

**Comment by poettering** [comment]
@yuwata ah, thanks for tracking this down! I fixed my patch accordingly now. Let's see if this works now.

(I also added two more, minor patches on top)

PTAL!

**Comment by poettering** [comment]
Wohoo. All CI passed, except:

```
dpkg: warning: while removing unattended-upgrades, directory '/var/log/unattended-upgrades' not empty so not removed
Processing triggers for ureadahead (0.100.0-20) ...
Processing triggers for systemd (234-2ubuntu12.1) ...
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
```

**Pull request merged**